src/Simplify.cpp

```diff
@@ -133,6 +133,38 @@ bool propagate_indeterminate_expression(Expr e0, Expr e1, Expr e2, Type t, Expr
            propagate_indeterminate_expression(e2, t, expr);
 }
 
+class ExprIsPure : public IRVisitor {
+    using IRVisitor::visit;
+
+    void visit(const Call *op) {
+        if (!op->is_pure()) {
+            result = false;
+        } else {
+            IRVisitor::visit(op);
+        }
+    }
+
+    void visit(const Load *op) {
+        if (!op->image.defined() && !op->param.defined()) {
+            // It's a load from an internal buffer, which could
+            // mutate.
```

Comment:
I think even output buffers could mutate as well, due to update definitions on the output of a pipeline...